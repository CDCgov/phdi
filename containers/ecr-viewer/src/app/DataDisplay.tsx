import React, { ReactNode, useEffect, useState } from "react";
import { Button } from "@trussworks/react-uswds";
import classNames from "classnames";

import { ToolTipElement } from "@/app/ToolTipElement";

export interface DisplayDataProps {
  title?: string;
  className?: string;
  toolTip?: string;
  value?: string | React.JSX.Element | React.JSX.Element[] | React.ReactNode;
  dividerLine?: boolean;
}

/**
 * Functional component for displaying data.
 * @param props - Props for the component.
 * @param props.item - The display data item to be rendered.
 * @param [props.className] - Additional class name for styling purposes.
 * @returns - A React element representing the display of data.
 */
export const DataDisplay: React.FC<{
  item: DisplayDataProps;
  className?: string;
}> = ({
  item,
  className,
}: {
  item: DisplayDataProps;
  className?: string;
}): React.JSX.Element => {
  item.dividerLine =
    item.dividerLine == null || item.dividerLine == undefined
      ? true
      : item.dividerLine;
  return (
    <div>
      <div className="grid-row">
        <div className="data-title">
          <ToolTipElement content={item.title} toolTip={item.toolTip} />
        </div>
        <div
          className={classNames(
            "grid-col-auto maxw7 text-pre-line",
            className,
            item.className ? item.className : "",
          )}
        >
          <FieldValue>{item.value}</FieldValue>
        </div>
      </div>
      {item.dividerLine ? <div className={"section__line_gray"} /> : ""}
    </div>
  );
};
/**
 * Functional component for displaying a value. If the value has a length greater than 500 characters, it will be split after 300 characters with a view more button to view the entire value.
 * @param value - props for the component
 * @param value.children - the value to be displayed in the value
 * @returns - A React element representing the display of the value
 */
const FieldValue: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  const maxLength = 500;
  const cutLength = 300;
  const [hidden, setHidden] = useState(true);
  const [fieldValue, setFieldValue] = useState(children);
  const valueLength = getReactNodeLength(children);
  const cutField = trimField(children, cutLength, setHidden).value;
  useEffect(() => {
    if (valueLength > maxLength) {
      if (hidden) {
        setFieldValue(cutField);
      } else {
        setFieldValue(
          <>
            {children}&nbsp;
            <Button
              type={"button"}
              unstyled={true}
              onClick={() => setHidden(true)}
            >
              View less
            </Button>
          </>,
        );
      }
    }
  }, [hidden]);

  return fieldValue;
};
/**
 * Recursively determine the character length of a ReactNode
 * @param value - react node to be measured
 * @returns - the number of characters in the ReactNode
 */
const getReactNodeLength = (value: React.ReactNode): number => {
  if (typeof value === "string") {
    return value.length;
  } else if (Array.isArray(value)) {
    let count = 0;
    value.forEach((val) => (count += getReactNodeLength(val)));
    return count;
  } else if (React.isValidElement(value) && value.props.children) {
    return getReactNodeLength(value.props.children);
  }
  return 0;
};
/**
 * Create an element with `remainingLength` length followed by a view more button
 * @param value - the value that will be cut
 * @param remainingLength - the length of how long the returned element will be
 * @param setHidden - a function used to signify that the view more button has been clicked.
 * @returns - an object with the shortened value and the length left over.
 */
const trimField = (
  value: React.ReactNode,
  remainingLength: number,
  setHidden: (val: boolean) => void,
): { value: React.ReactNode; remainingLength: number } => {
  if (remainingLength < 1) {
    return { value: null, remainingLength };
  }
  if (typeof value === "string") {
    const cutString = value.substring(0, remainingLength);
    if (remainingLength - cutString.length === 0) {
      return {
        value: (
          <>
            {cutString}...&nbsp;
            <Button
              type={"button"}
              unstyled={true}
              onClick={() => setHidden(false)}
            >
              View more
            </Button>
          </>
        ),
        remainingLength: 0,
      };
    }
    return {
      value: cutString,
      remainingLength: remainingLength - cutString.length,
    };
  } else if (Array.isArray(value)) {
    let newValArr = [];
    for (let i = 0; i < value.length; i++) {
      let splitVal = trimField(value[i], remainingLength, setHidden);
      remainingLength = splitVal.remainingLength;
      newValArr.push(
        <React.Fragment key={`arr-${i}-${splitVal.value}`}>
          {splitVal.value}
        </React.Fragment>,
      );
    }
    return { value: newValArr, remainingLength: remainingLength };
  } else if (React.isValidElement(value) && value.props.children) {
    let childrenCopy: ReactNode;
    if (Array.isArray(value.props.children)) {
      childrenCopy = [...value.props.children];
    } else {
      childrenCopy = value.props.children;
    }
    let split = trimField(childrenCopy, remainingLength, setHidden);
    const newElement = React.cloneElement(
      value,
      { ...value.props },
      split.value,
    );
    return { value: newElement, remainingLength: split.remainingLength };
  }
  return { value, remainingLength: remainingLength };
};
