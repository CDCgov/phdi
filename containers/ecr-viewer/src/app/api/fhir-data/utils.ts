import * as fs from 'fs';
import * as path from 'path';
import yaml from 'js-yaml';

export function loadYamlConfig() {
  const filePath = path.join(process.cwd(), 'src/app/api/fhir-data/fhirPath.yml');
  const fileContents = fs.readFileSync(filePath, 'utf8');
  const data = yaml.load(fileContents);
  return data;
}
