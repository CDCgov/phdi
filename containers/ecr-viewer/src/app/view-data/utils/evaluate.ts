import {
  evaluate as fhirPathEvaluate,
  type Model,
  type Path,
  type UserInvocationTable,
} from "fhirpath";
import { Context } from "node:vm";

let cache: { [key: string]: any } = {};

/**
 * Evaluates a FHIRPath expression on the provided FHIR data.
 * @param fhirData - The FHIR data to evaluate the FHIRPath expression on.
 * @param path - The FHIRPath expression to evaluate.
 * @param [context] - Optional context object to provide additional data for evaluation.
 * @param [model] - Optional model object to provide additional data for evaluation.
 * @param [options] - Optional options object for additional configuration.
 * @param [options.resolveInternalTypes] - Whether to resolve internal types in the evaluation.
 * @param [options.traceFn] - Optional trace function for logging evaluation traces.
 * @param [options.userInvocationTable] - Optional table for tracking user invocations.
 * @returns - An array containing the result of the evaluation.
 */
export const evaluate = (
  fhirData: any,
  path: string | Path,
  context?: Context,
  model?: Model,
  options?: {
    resolveInternalTypes?: boolean;
    traceFn?: (value: any, label: string) => void;
    userInvocationTable?: UserInvocationTable;
  },
): any[] => {
  let key =
    JSON.stringify(fhirData) +
    ",,,,," +
    JSON.stringify(context) +
    ",,,,," +
    JSON.stringify(path);
  if (cache.hasOwnProperty(key)) {
    return cache[key];
  }
  cache[key] = fhirPathEvaluate(fhirData, path, context, model, options);
  return cache[key];
};
