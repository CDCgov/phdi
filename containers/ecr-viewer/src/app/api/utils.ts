import * as fs from "fs";
import * as path from "path";
import yaml from "js-yaml";
import { PathMappings } from "@/app/view-data/utils/utils";

/**
 * Loads the YAML configuration for path mappings from a predefined file location.
 * @returns An object representing the path mappings defined in the YAML configuration file.
 */
export function loadYamlConfig(): PathMappings {
  const filePath = path.join(process.cwd(), "src/app/api/fhirPath.yml");
  const fileContents = fs.readFileSync(filePath, "utf8");
  return <PathMappings>yaml.load(fileContents);
}

/**
 * Converts stream data to json data
 * @param stream - The input stream that provides JSON data in chunks. The stream
 *   should implement the async iterable protocol to allow for-await-of
 *   iteration over its data chunks.
 * @returns A promise that resolves to the JSON-parsed object from the accumulated
 *  stream data. The specific structure of this object depends on the JSON
 *  content of the stream.
 */
export async function streamToJson(stream: any) {
  let rawData = "";
  for await (const chunk of stream) {
    rawData += chunk;
  }
  return JSON.parse(rawData);
}
