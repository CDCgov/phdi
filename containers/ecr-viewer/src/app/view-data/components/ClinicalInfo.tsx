import { DataDisplay, DisplayData, DataTableDisplay } from "@/app/utils";
import {
  AccordianSection,
  AccordianH4,
  AccordianDiv,
} from "../component-utils";
import { SectionConfig } from "./SideNav";
import React from "react";

interface ClinicalProps {
  reasonForVisitDetails: DisplayData[];
  activeProblemsDetails: DisplayData[];
  vitalData: DisplayData[];
  immunizationsDetails: DisplayData[];
  treatmentData: DisplayData[];
  clinicalNotes: DisplayData[];
}

export const clinicalInfoConfig: SectionConfig = new SectionConfig(
  "Clinical Info",
  [
    "Symptoms and Problems",
    "Immunizations",
    "Diagnostics and Vital Signs",
    "Treatment Details",
    "Clinical Notes",
  ],
);

export const ClinicalInfo = ({
  reasonForVisitDetails,
  activeProblemsDetails,
  immunizationsDetails,
  vitalData,
  treatmentData,
  clinicalNotes,
}: ClinicalProps) => {
  const renderTableDetails = (tableDetails: DisplayData[]) => {
    return (
      <div>
        {tableDetails.map((item, index) => (
          <div key={index}>
            <div className="grid-col-auto text-pre-line">{item.value}</div>
            <div className={"section__line_gray"} />
          </div>
        ))}
      </div>
    );
  };

  const renderClinicalNotes = () => {
    return (
      <>
        <AccordianH4>
          <span id={clinicalInfoConfig.subNavItems?.[4].id}>
            {clinicalInfoConfig.subNavItems?.[4].title}
          </span>
        </AccordianH4>
        <AccordianDiv className={"clinical_info_container"}>
          {clinicalNotes.map((item, index) => {
            let className = "";
            if (
              React.isValidElement(item.value) &&
              item.value.type == "table"
            ) {
              className = "maxw-full grid-col-12 margin-top-1";
            }
            return (
              <DataDisplay item={item} key={index} className={className} />
            );
          })}
        </AccordianDiv>
      </>
    );
  };

  const renderSymptomsAndProblems = () => {
    return (
      <>
        <AccordianH4>
          <span id={clinicalInfoConfig.subNavItems?.[0].id}>
            {clinicalInfoConfig.subNavItems?.[0].title}
          </span>
        </AccordianH4>
        <AccordianDiv>
          <div data-testid="reason-for-visit">
            {reasonForVisitDetails.map((item, index) => (
              <DataDisplay item={item} key={index} />
            ))}
          </div>
          <div data-testid="active-problems">
            {renderTableDetails(activeProblemsDetails)}
          </div>
        </AccordianDiv>
      </>
    );
  };

  const renderImmunizationsDetails = () => {
    return (
      <>
        <AccordianH4>
          <span id={clinicalInfoConfig.subNavItems?.[1].id}>
            {clinicalInfoConfig.subNavItems?.[1].title}
          </span>
        </AccordianH4>
        <AccordianDiv>
          <div data-testid="immunization-history">
            {renderTableDetails(immunizationsDetails)}
          </div>
        </AccordianDiv>
      </>
    );
  };

  const renderVitalDetails = () => {
    return (
      <>
        <AccordianH4>
          <span id={clinicalInfoConfig.subNavItems?.[2].id}>
            {clinicalInfoConfig.subNavItems?.[2].title}
          </span>
        </AccordianH4>
        <AccordianDiv>
          <div className="lh-18" data-testid="vital-signs">
            {vitalData.map((item, index) => (
              <DataDisplay item={item} key={index} />
            ))}
          </div>
        </AccordianDiv>
      </>
    );
  };

  // TODO: HERE
  const renderTreatmentDetails = () => {
    const data = treatmentData.filter((item) => !React.isValidElement(item));
    return (
      <>
        <AccordianH4>
          <span id={clinicalInfoConfig.subNavItems?.[3].id}>
            {clinicalInfoConfig.subNavItems?.[3].title}
          </span>
        </AccordianH4>
        <AccordianDiv>
          <div data-testid="treatment-details">
            {data.map((item, index) => (
              <DataTableDisplay item={item} key={index} />
            ))}
          </div>
          <div className={"section__line_gray margin-y-2"} />
        </AccordianDiv>
      </>
    );
  };

  return (
    <AccordianSection>
      {clinicalNotes?.length > 0 && renderClinicalNotes()}
      {(reasonForVisitDetails.length > 0 || activeProblemsDetails.length > 0) &&
        renderSymptomsAndProblems()}
      {treatmentData.length > 0 && renderTreatmentDetails()}
      {immunizationsDetails.length > 0 && renderImmunizationsDetails()}
      {vitalData.length > 0 && renderVitalDetails()}
    </AccordianSection>
  );
};

export default ClinicalInfo;
