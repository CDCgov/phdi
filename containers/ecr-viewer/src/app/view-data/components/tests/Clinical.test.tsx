import React from "react";
import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import ClinicalInfo, { TableEntry } from "../ClinicalInfo";
import { loadYamlConfig } from "@/app/api/utils";
import { returnProceduresTable, evaluateClinicalData } from "@/app/utils";
import { Procedure } from "fhir/r4";

describe("Snapshot test for Vital Signs/Encounter (Clinical Info section)", () => {
  let container: HTMLElement;

  beforeAll(() => {
    const proceduresArray = [
      {
        id: "b40f0081-4052-4971-3f3b-e3d9f5e1e44d",
        code: {
          coding: [
            {
              code: "0241U",
              system: "http://www.ama-assn.org/go/cpt",
              display:
                "HC INFECTIOUS DISEASE PATHOGEN SPECIFIC RNA SARS-COV-2/INF A&B/RSV UPPER RESP SPEC DETECTED OR NOT",
            },
          ],
        },
        meta: {
          source: ["ecr"],
          profile: [
            "http://hl7.org/fhir/us/core/StructureDefinition/us-core-procedure",
          ],
        },
        reason: [
          {
            display: "Struck by nonvenomous lizards, sequela",
            reference: "e436d9b7-6b4e-f553-0314-5a388d15e02e",
          },
        ],
        status: "completed",
        subject: {
          reference: "Patient/5360b569-1354-4ece-b6a1-58b0946fc861",
        },
        identifier: [
          {
            value: "2884257^",
            system: "urn:oid:1.2.840.114350.1.13.502.3.7.1.1988.1",
          },
        ],
        resourceType: "Procedure",
        performedDateTime: "06/24/2022",
      },
      {
        id: "b40f0081-4052-4971-3f3b-e3d9f5e1e44d",
        code: {
          coding: [
            {
              code: "86308",
              system: "http://www.ama-assn.org/go/cpt",
              display: "HC HETEROPHILE ANTIBODIES SCREENING",
            },
          ],
        },
        meta: {
          source: ["ecr"],
          profile: [
            "http://hl7.org/fhir/us/core/StructureDefinition/us-core-procedure",
          ],
        },
        reason: [
          {
            display:
              "Routine general medical examination at a health care facility",
            reference: "7cda2e3e-5d91-428f-8abe-517846d4749e",
          },
        ],
        status: "completed",
        subject: {
          reference: "Patient/5360b569-1354-4ece-b6a1-58b0946fc861",
        },
        identifier: [
          {
            value: "2884257^",
            system: "urn:oid:1.2.840.114350.1.13.502.3.7.1.1988.1",
          },
        ],
        resourceType: "Procedure",
        performedDateTime: "06/16/2022",
      },
    ] as unknown as Procedure[];
    const mappings = {
      procedures: "Bundle.entry.resource.where(resourceType='Procedure')",
      procedureName: "Procedure.code.coding.first().display",
      procedureDate: "Procedure.performedDateTime",
      procedureReason: "Procedure.reason.display",
    };
    const treatmentData = [
      {
        title: "Procedures",
        value: returnProceduresTable(proceduresArray, mappings),
      },
    ];
    const vitalData = [
      {
        title: "Vitals",
        value: `Height: 65 inches\n\nWeight: 150 Lbs\n\nBody Mass Index (BMI): 25`,
      },
      {
        title: "Facility Name",
        value: "PRM- Palmdale Regional Medical Center",
      },
      {
        title: "Facility Type",
        value: "Healthcare Provider",
      },
    ];
    container = render(
      <ClinicalInfo
        clinicalNotes={[]}
        activeProblemsDetails={[]}
        vitalData={vitalData}
        reasonForVisitDetails={[]}
        immunizationsDetails={[]}
        treatmentData={treatmentData}
        planOfTreatment={[]}
      />,
    ).container;
  });
  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });
  it("should pass accessibility test", async () => {
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Snapshot test for Clinical Notes", () => {
  it("should match snapshot for non table notes", async () => {
    const clinicalNotes = [
      {
        title: "Miscellaneous Notes",
        value: (
          <p>
            This patient was only recently discharged for a recurrent GI bleed
            as described
          </p>
        ),
      },
    ];
    let { container } = render(
      <ClinicalInfo
        clinicalNotes={clinicalNotes}
        activeProblemsDetails={[]}
        vitalData={[]}
        reasonForVisitDetails={[]}
        immunizationsDetails={[]}
        treatmentData={[]}
        planOfTreatment={[]}
      />,
    );
    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });
  it("should match snapshot for table notes", async () => {
    const clinicalNotes = [
      {
        title: "Miscellaneous Notes",
        value: (
          <table>
            <thead>
              <tr>
                <th>Active Problems</th>
                <th>Noted Date</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>Parkinson's syndrome</td>
                <td>7/25/22</td>
              </tr>
              <tr>
                <td>Essential hypertension</td>
                <td>7/21/22</td>
              </tr>
            </tbody>
          </table>
        ),
      },
    ];
    let { container } = render(
      <ClinicalInfo
        clinicalNotes={clinicalNotes}
        activeProblemsDetails={[]}
        vitalData={[]}
        reasonForVisitDetails={[]}
        immunizationsDetails={[]}
        treatmentData={[]}
        planOfTreatment={[]}
      />,
    );
    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Check that Clinical Info components render given FHIR bundle", () => {
  const fhirBundleClinicalInfo = require("../../../tests/assets/BundleClinicalInfo.json");
  const mappings = loadYamlConfig();
  const testClinicalData = evaluateClinicalData(
    fhirBundleClinicalInfo,
    mappings,
  );

  const testImmunizationsData =
    testClinicalData.immunizationsDetails.availableData;
  const testActiveProblemsData =
    testClinicalData.activeProblemsDetails.availableData;
  const testVitalSignsData = testClinicalData.vitalData.availableData;
  const testReasonForVisitData =
    testClinicalData.reasonForVisitDetails.availableData;
  const testTreatmentData = testClinicalData.treatmentData.availableData;

  const testPlanOfTreatment: {
    value: TableEntry[] | undefined;
  }[] = [
    {
      value: [
        {
          Name: "PCR SARS-CoV-2 and Influenza A/B",
          Type: "Lab",
          Priority: "Routine",
          AssociatedDiagnoses: "",
          DateTime: "12/07/2021 4:16 PM CST",
          OrderSchedule: "",
        },
        {
          Name: "Drugs Of Abuse Comprehensive Screen, Ur",
          Type: "Lab",
          Priority: "STAT",
          AssociatedDiagnoses: "",
          DateTime: "12/23/2022 11:13 AM PST",
          OrderSchedule: "",
        },
      ],
    },
  ];

  test("eCR Viewer renders immunization table given FHIR bundle with immunization info", () => {
    const clinicalInfo = render(
      <ClinicalInfo
        immunizationsDetails={testImmunizationsData}
        reasonForVisitDetails={[]}
        activeProblemsDetails={[]}
        vitalData={[]}
        treatmentData={[]}
        clinicalNotes={[]}
        planOfTreatment={[]}
      />,
    );

    // Ensure that Immunizations Section renders
    const expectedImmunizationsElement = clinicalInfo.getByTestId(
      "immunization-history",
    );
    expect(expectedImmunizationsElement).toBeInTheDocument();

    // Ensure only one table (Immunization History) is rendering
    const expectedTable = clinicalInfo.getAllByTestId("table");
    expect(expectedTable[0]).toBeInTheDocument();
    expect(expectedTable.length).toEqual(1);
  });

  test("eCR Viewer renders active problems table given FHIR bundle with active problems info", () => {
    const clinicalInfo = render(
      <ClinicalInfo
        immunizationsDetails={[]}
        reasonForVisitDetails={[]}
        activeProblemsDetails={testActiveProblemsData}
        vitalData={[]}
        treatmentData={[]}
        clinicalNotes={[]}
        planOfTreatment={[]}
      />,
    );

    const expectedActiveProblemsElement =
      clinicalInfo.getByTestId("active-problems");
    expect(expectedActiveProblemsElement).toBeInTheDocument();

    // Ensure only one table (Active Problems) is rendering
    const expectedTable = clinicalInfo.getAllByTestId("table");
    expect(expectedTable[0]).toBeInTheDocument();
    expect(expectedTable.length).toEqual(1);
  });

  test("eCR Viewer renders vital signs given FHIR bundle with vital signs info", () => {
    const clinicalInfo = render(
      <ClinicalInfo
        immunizationsDetails={[]}
        reasonForVisitDetails={[]}
        activeProblemsDetails={[]}
        vitalData={testVitalSignsData}
        treatmentData={[]}
        clinicalNotes={[]}
        planOfTreatment={[]}
      />,
    );

    const expectedVitalSignsElement = clinicalInfo.getByTestId("vital-signs");
    expect(expectedVitalSignsElement).toBeInTheDocument();
  });

  test("eCR Viewer renders reason for visit given FHIR bundle with reason for visit info", () => {
    const clinicalInfo = render(
      <ClinicalInfo
        immunizationsDetails={[]}
        reasonForVisitDetails={testReasonForVisitData}
        activeProblemsDetails={[]}
        vitalData={[]}
        treatmentData={[]}
        clinicalNotes={[]}
        planOfTreatment={[]}
      />,
    );

    const expectedReasonForVisitElement =
      clinicalInfo.getByTestId("reason-for-visit");
    expect(expectedReasonForVisitElement).toBeInTheDocument();
  });

  test("eCR Viewer renders treatment data given FHIR bundle with treatment data info", () => {
    const clinicalInfo = render(
      <ClinicalInfo
        immunizationsDetails={[]}
        reasonForVisitDetails={[]}
        activeProblemsDetails={[]}
        vitalData={[]}
        treatmentData={testTreatmentData}
        clinicalNotes={[]}
        planOfTreatment={[]}
      />,
    );

    const expectedTreatmentElement =
      clinicalInfo.getByTestId("treatment-details");
    expect(expectedTreatmentElement).toBeInTheDocument();

    // Ensure only one table (Treatment) is rendering
    const expectedTable = clinicalInfo.getAllByTestId("table");
    expect(expectedTable[0]).toBeInTheDocument();
    expect(expectedTable.length).toEqual(1);
  });

  test("eCR Viewer renders treatment data given plan of treatment", () => {
    const clinicalInfo = render(
      <ClinicalInfo
        immunizationsDetails={[]}
        reasonForVisitDetails={[]}
        activeProblemsDetails={[]}
        vitalData={[]}
        treatmentData={[]}
        clinicalNotes={[]}
        planOfTreatment={testPlanOfTreatment}
      />,
    );

    const expectedTreatmentElement = clinicalInfo.getByText("Pending Results");
    expect(expectedTreatmentElement).toBeInTheDocument();

    // Ensure only one table (Treatment) is rendering
    const expectedTable = clinicalInfo.getAllByTestId("table");
    expect(expectedTable[0]).toBeInTheDocument();
    expect(expectedTable.length).toEqual(1);
  });

  test("eCR Viewer renders all Clinical Info sections", () => {
    const clinicalInfo = render(
      <ClinicalInfo
        immunizationsDetails={testImmunizationsData}
        reasonForVisitDetails={testReasonForVisitData}
        activeProblemsDetails={testActiveProblemsData}
        vitalData={testVitalSignsData}
        treatmentData={testTreatmentData}
        clinicalNotes={[]}
        planOfTreatment={testPlanOfTreatment}
      />,
    );

    const expectedImmunizationsElement = clinicalInfo.getByTestId(
      "immunization-history",
    );
    expect(expectedImmunizationsElement).toBeInTheDocument();

    const expectedActiveProblemsElement =
      clinicalInfo.getByTestId("active-problems");
    expect(expectedActiveProblemsElement).toBeInTheDocument();

    const expectedTreatmentElement =
      clinicalInfo.getByTestId("treatment-details");
    expect(expectedTreatmentElement).toBeInTheDocument();

    // Ensure the three tables (Immunization History & Active Problems & Treatment) are rendering
    const expectedTable = clinicalInfo.getAllByTestId("table");
    expect(expectedTable[0]).toBeInTheDocument();
    expect(expectedTable.length).toEqual(4);

    const expectedVitalSignsElement = clinicalInfo.getByTestId("vital-signs");
    expect(expectedVitalSignsElement).toBeInTheDocument();

    const expectedReasonForVisitElement =
      clinicalInfo.getByTestId("reason-for-visit");
    expect(expectedReasonForVisitElement).toBeInTheDocument();
  });
});
