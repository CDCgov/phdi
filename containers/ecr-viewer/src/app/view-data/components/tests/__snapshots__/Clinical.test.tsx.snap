// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for Clinical Notes should match snapshot for non table notes 1`] = `
<div>
  <div
    class="margin-top-0"
  >
    <div
      class="padding-bottom-3"
    >
      <div
        class="usa-summary-box__body"
      >
        <h4
          class="usa-summary-box__heading padding-y-105"
          id="summary-box-key-information"
        >
          <span
            id="clinical-notes"
          >
            Clinical Notes
          </span>
        </h4>
        <div
          class="usa-summary-box__text clinical_info_container"
        >
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title"
              >
                Miscellaneous Notes
              </div>
              <div
                class="grid-col-auto maxw7 text-pre-line"
              >
                <p>
                  This patient was only recently discharged for a recurrent GI bleed as described
                </p>
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot test for Clinical Notes should match snapshot for table notes 1`] = `
<div>
  <div
    class="margin-top-0"
  >
    <div
      class="padding-bottom-3"
    >
      <div
        class="usa-summary-box__body"
      >
        <h4
          class="usa-summary-box__heading padding-y-105"
          id="summary-box-key-information"
        >
          <span
            id="clinical-notes"
          >
            Clinical Notes
          </span>
        </h4>
        <div
          class="usa-summary-box__text clinical_info_container"
        >
          <div>
            <div>
              <div
                class="grid-col-auto text-pre-line"
              >
                <table>
                  <caption>
                    Miscellaneous Notes
                  </caption>
                  <thead>
                    <tr>
                      <th>
                        Active Problems
                      </th>
                      <th>
                        Noted Date
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>
                        Parkinson's syndrome
                      </td>
                      <td>
                        7/25/22
                      </td>
                    </tr>
                    <tr>
                      <td>
                        Essential hypertension
                      </td>
                      <td>
                        7/21/22
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot test for Vital Signs/Encounter (Clinical Info section) should match snapshot 1`] = `
<div>
  <div
    class="margin-top-0"
  >
    <div
      class="padding-bottom-3"
    >
      <div
        class="usa-summary-box__body"
      >
        <h4
          class="usa-summary-box__heading padding-y-105"
          id="summary-box-key-information"
        >
          <span
            id="treatment-details"
          >
            Treatment Details
          </span>
        </h4>
        <div
          class="usa-summary-box__text"
        >
          <div
            data-testid="treatment-details"
          >
            <div
              class="grid-row"
            >
              <div
                class="grid-col-auto width-full text-pre-line"
              >
                <table
                  class="usa-table usa-table--borderless width-full src-components-Table-Table-module__fixed--Eq00W table-caption-margin margin-y-0 border-top border-left border-right"
                  data-testid="table"
                >
                  <caption>
                    Procedures
                  </caption>
                  <thead>
                    <tr>
                      <th
                        class="tableHeader"
                        scope="col"
                      >
                        Name
                      </th>
                      <th
                        class="tableHeader"
                        scope="col"
                      >
                        Date Performed
                      </th>
                      <th
                        class="tableHeader"
                        scope="col"
                      >
                        Reason
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td
                        class="text-top"
                      >
                        HC INFECTIOUS DISEASE PATHOGEN SPECIFIC RNA SARS-COV-2/INF A&B/RSV UPPER RESP SPEC DETECTED OR NOT
                      </td>
                      <td
                        class="text-top"
                      >
                        06/24/2022
                      </td>
                      <td
                        class="text-top"
                      >
                        Struck by nonvenomous lizards, sequela
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="text-top"
                      >
                        HC HETEROPHILE ANTIBODIES SCREENING
                      </td>
                      <td
                        class="text-top"
                      >
                        06/16/2022
                      </td>
                      <td
                        class="text-top"
                      >
                        Routine general medical examination at a health care facility
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
          </div>
          <div
            class="section__line_gray margin-y-2"
          />
        </div>
        <h4
          class="usa-summary-box__heading padding-y-105"
          id="summary-box-key-information"
        >
          <span
            id="diagnostics-and-vital-signs"
          >
            Diagnostics and Vital Signs
          </span>
        </h4>
        <div
          class="usa-summary-box__text"
        >
          <div
            class="lh-18"
            data-testid="vital-signs"
          >
            <div>
              <div
                class="grid-row"
              >
                <div
                  class="data-title"
                >
                  Vitals
                </div>
                <div
                  class="grid-col-auto maxw7 text-pre-line"
                >
                  Height: 65 inches

Weight: 150 Lbs

Body Mass Index (BMI): 25
                </div>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
            <div>
              <div
                class="grid-row"
              >
                <div
                  class="data-title"
                >
                  Facility Name
                </div>
                <div
                  class="grid-col-auto maxw7 text-pre-line"
                >
                  PRM- Palmdale Regional Medical Center
                </div>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
            <div>
              <div
                class="grid-row"
              >
                <div
                  class="data-title"
                >
                  Facility Type
                </div>
                <div
                  class="grid-col-auto maxw7 text-pre-line"
                >
                  Healthcare Provider
                </div>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
